server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -484,18 +483,27 @@ public ShardRouting reinitializeReplicaShard() {
         );
     }
 
+    public ShardRouting moveToStarted() {
+        return moveToStarted(null);
+    }
+
     /**
      * Set the shards state to <code>STARTED</code>. The shards state must be
      * <code>INITIALIZING</code> or <code>RELOCATING</code>. Any relocation will be
      * canceled.
      */
-    public ShardRouting moveToStarted() {
+    public ShardRouting moveToStarted(@Nullable Metadata metadata) {
```

Comment:
I think we should have an extra field within `ShardRouting` to tell whether or not this shard is a searchable snapshot, that should simplify things and allows us to have clearer assertions. wdyt?

Reply:
I think it would add clarity and simplify this where we now have more API surface, i.e. why do we have 2 `moveToStarted` methods? additionally we could add more assertions to ensure that we only populate the expected shard size when it's needed.

Reply:
WDYT about addressing this in a separate PR? `ShardRouting`'s constructor gets called from a lot of places, so adding a new field would add a significant amount of changes in the classes unrelated to the PR.

Reply:
I would also like to get rid of the extra `moveToStarted` variant. Being lenient and assuming null in this central place seems dangerous.

But I do see the problems in doing so, at least in the test code, where we often do not have `Metadata` nor `IndexMetadata`. I wonder if we could fix this by instead letting `moveToStarted` take `expectedShardSize` as parameter. We can then let `AllocationService derive the `expectedShardSize` from the metadata it already has available.

I think this would help separate responsibilities too.

All tests that do not care can then pass in `UNAVAILABLE_EXPECTED_SHARD_SIZE`.